src/vsphere_cpi/spec/integration/vm_host_rule_spec.rb

```diff
@@ -0,0 +1,402 @@
+          expect(get_count_vm_host_affinity_rules(cluster_mob)).to eq(0)
+        end
+        context 'when all hosts in first cluster host group are in maintenance mode' do
+          before do
+            turn_maintenance_on_for_hosts(cpi, first_host_group_mob.host)
+          end
+          after do
+            turn_maintenance_off_for_hosts(cpi, first_host_group_mob.host)
+          end
+          it 'should create one VM on one of the hosts in the host groups of second cluster' do
+            simple_vm_lifecycle(cpi, '', vm_type, get_network_spec) do |vm_id|
+              vm = cpi.vm_provider.find(vm_id)
+              vm_host = vm.mob.runtime.host.name
+              expect(third_host_group_mob.host.map(&:name)).to include(vm_host)
+              expect(vm.cluster).to eq(@second_cluster_name)
+            end
+            expect(get_count_vm_host_affinity_rules(cluster_mob) +
+              get_count_vm_host_affinity_rules(second_cluster_mob)).to eq(0)
+          end
+          it 'should create and delete multiple VMs (5) in parallel' do
+            thread_list = []
+            5.times do
+              thread_list << Thread.new do
+                simple_vm_lifecycle(cpi, '', vm_type, get_network_spec) do |vm_id|
+                  vm = cpi.vm_provider.find(vm_id)
+                  vm_host = vm.mob.runtime.host.name
+                  expect(third_host_group_mob.host.map(&:name)).to include(vm_host)
+                  expect(vm.cluster).to eq(@second_cluster_name)
+                end
+              end
+            end
+            thread_list.each {|thread| thread.join}
+            expect(get_count_vm_host_affinity_rules(cluster_mob)).to eq(0)
+          end
+        end
+      end
+      context 'when one of the cluster has host group and other has resource pool' do
+        let(:options) do
+          cpi_options(
+            datacenters: [{
+              'datastore_pattern' => @datastore_shared_pattern,
+              'persistent_datastore_pattern' => @datastore_shared_pattern,
+              clusters: [
+                { @cluster_name => { host_group: @second_host_group, } },
+                { @second_cluster_name => { resource_pool: @second_cluster_resource_pool_name, } },
+              ]
+            }]
+          )
+        end
+        let(:cpi) do
+          VSphereCloud::Cloud.new(options)
+        end
+        it 'should create one VM on the second cluster without host group\
+            (as resource pool has more memory available than host group)' do
+          simple_vm_lifecycle(cpi, '', vm_type, get_network_spec) do |vm_id|
+            vm = cpi.vm_provider.find(vm_id)
+            expect(vm.cluster).to eq(@second_cluster_name)
+          end
+          expect(get_count_vm_host_affinity_rules(cluster_mob) +
+            get_count_vm_host_affinity_rules(second_cluster_mob)).to eq(0)
+        end
+        it 'should create multiple VMs (3) one after another on the second cluster\'s resource pool' do
```

Comment:
creation of multiple VM's for different configurations is repetitive and I think unnecessary for all the context in this spec file

Reply:
Letting this stay in favor of repeated testing for any non-deterministic bug.
Will take it our after the release is cut and feature stabilizes.